src/Symfony/Component/Message/MessageBus.php

```diff
@@ -0,0 +1,53 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Message;
+
+/**
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ * @author Matthias Noback <matthiasnoback@gmail.com>
+ */
+class MessageBus implements MessageBusInterface
+{
+    /**
+     * @var MiddlewareInterface[]
+     */
+    private $middlewares;
+
+    /**
+     * @param MiddlewareInterface[]|iterable $middlewares
+     */
+    public function __construct(iterable $middlewares = array())
+    {
+        $this->middlewares = is_array($middlewares) ? array_values($middlewares) : iterator_to_array($middlewares, false);
```

Comment:
no laziness for middlewares? Can't they be heavy to instantiate?

Reply:
Not sure it matters. All middlewares are instantiated (one after the other, but all are required for the message handling). Doing this in the constructor might be avoided, though, so they are not instantiated just by injecting the bus.

Reply:
i suggested this as we rely on indexes..

Reply:
Nothing wrong with it. But moving it to be executed only on first dispatch would avoid instantiating the middlewares even if the bus happens to be injected somewhere but not used at runtime. Probably small gain, but still.

Reply:
Added the micro-optimisation in df2a59c077 :)